File: src/main/java/com/jaredrobertson/plugins/angularFileSwitcher/settings/AppSettingsConfigurable.java
// Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package com.jaredrobertson.plugins.angularFileSwitcher.settings;

import com.intellij.openapi.options.Configurable;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

/**
 * Provides controller functionality for application settings.
 */
public class AppSettingsConfigurable implements Configurable {
    private AppSettingsComponent mySettingsComponent;

    @Nls(capitalization = Nls.Capitalization.Title)
    @Override
    public String getDisplayName() {
        return "Angular File Switcher";
    }

    @Override
    public JComponent getPreferredFocusedComponent() {
        return mySettingsComponent.getPreferredFocusedComponent();
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        mySettingsComponent = new AppSettingsComponent();
        return mySettingsComponent.getPanel();
    }

    @Override
    public boolean isModified() {
        AppSettingsState settings = AppSettingsState.getInstance();
        boolean modified = !mySettingsComponent.getClassFileExtensionsText().equals(settings.classFileExtensions);
        modified |= !mySettingsComponent.getTemplateFileExtensionsText().equals(settings.templateFileExtensions);
        modified |= !mySettingsComponent.getStyleFileExtensionsText().equals(settings.styleFileExtensions);
        modified |= !mySettingsComponent.getTestFileExtensionsText().equals(settings.testFileExtensions);
        modified |= mySettingsComponent.getSwitcherGrouping() != settings.switcherGrouping;
        modified |= mySettingsComponent.getCloseBehavior() != settings.closeBehavior;
        return modified;
    }

    @Override
    public void apply() {
        AppSettingsState settings = AppSettingsState.getInstance();
        settings.classFileExtensions = mySettingsComponent.getClassFileExtensionsText();
        settings.templateFileExtensions = mySettingsComponent.getTemplateFileExtensionsText();
        settings.styleFileExtensions = mySettingsComponent.getStyleFileExtensionsText();
        settings.testFileExtensions = mySettingsComponent.getTestFileExtensionsText();
        settings.switcherGrouping = mySettingsComponent.getSwitcherGrouping();
        settings.closeBehavior = mySettingsComponent.getCloseBehavior();
    }

    @Override
    public void reset() {
        AppSettingsState settings = AppSettingsState.getInstance();
        mySettingsComponent.setClassFileExtensionsText(settings.classFileExtensions);
        mySettingsComponent.setTemplateFileExtensionsText(settings.templateFileExtensions);
        mySettingsComponent.setStyleFileExtensionsText(settings.styleFileExtensions);
        mySettingsComponent.setTestFileExtensionsText(settings.testFileExtensions);
        mySettingsComponent.setSwitcherGrouping(settings.switcherGrouping);
        mySettingsComponent.setCloseBehavior(settings.closeBehavior);
    }

    @Override
    public void disposeUIResources() {
        mySettingsComponent = null;
    }

}

File: src/main/java/com/jaredrobertson/plugins/angularFileSwitcher/settings/AppSettingsConfigurable.kt
// Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.jaredrobertson.plugins.angularFileSwitcher.settings

import com.intellij.openapi.options.Configurable
import com.jaredrobertson.plugins.angularFileSwitcher.settings.AppSettingsState.Companion.instance
import org.jetbrains.annotations.Nls
import javax.swing.JComponent

/**
 * Provides controller functionality for application settings.
 */
class AppSettingsConfigurable : Configurable {
    private var mySettingsComponent: AppSettingsComponent? = null
    override fun getDisplayName(): @Nls(capitalization = Nls.Capitalization.Title) String {
        return "Angular File Switcher"
    }

    override fun getPreferredFocusedComponent(): JComponent {
        return mySettingsComponent!!.preferredFocusedComponent
    }

    override fun createComponent(): JComponent {
        mySettingsComponent = AppSettingsComponent()
        return mySettingsComponent!!.panel
    }

    override fun isModified(): Boolean {
        val settings = instance
        var modified = mySettingsComponent!!.classFileExtensionsText != settings.classFileExtensions
        modified =
            modified or (mySettingsComponent!!.templateFileExtensionsText != settings.templateFileExtensions)
        modified =
            modified or (mySettingsComponent!!.styleFileExtensionsText != settings.styleFileExtensions)
        modified =
            modified or (mySettingsComponent!!.testFileExtensionsText != settings.testFileExtensions)
        modified = modified or (mySettingsComponent!!.switcherGrouping !== settings.grouping)
        modified = modified or (mySettingsComponent!!.closeBehavior !== settings.closeBehavior)
        
        return modified
    }

    override fun apply() {
        val settings = instance
        settings.classFileExtensions = mySettingsComponent!!.classFileExtensionsText
        settings.templateFileExtensions = mySettingsComponent!!.templateFileExtensionsText
        settings.styleFileExtensions = mySettingsComponent!!.styleFileExtensionsText
        settings.testFileExtensions = mySettingsComponent!!.testFileExtensionsText
        settings.grouping = mySettingsComponent!!.switcherGrouping
        settings.closeBehavior = mySettingsComponent!!.closeBehavior
    }

    override fun reset() {
        val settings = instance
        mySettingsComponent!!.classFileExtensionsText = settings.classFileExtensions
        mySettingsComponent!!.templateFileExtensionsText = settings.templateFileExtensions
        mySettingsComponent!!.styleFileExtensionsText = settings.styleFileExtensions
        mySettingsComponent!!.testFileExtensionsText = settings.testFileExtensions
        mySettingsComponent!!.switcherGrouping = settings.grouping
        mySettingsComponent!!.closeBehavior = settings.closeBehavior
    }

    override fun disposeUIResources() {
        mySettingsComponent = null
    }
}